File: python/Minimum_Cost_to_Convert_String_II_2977.py
class Solution:

    class Trie:

        counter = 0

        def __init__(self):
            self.child = [None] * 26
            self.id = -1
        
        def insert(self, curStr):
            cur = self
            for curChar in curStr:
                curId = ord(curChar) - ord('a')
                if not cur.child[curId]:
                    cur.child[curId] = Solution.Trie()
                cur = cur.child[curId]
            if cur.id == -1:
                cur.id = Solution.Trie.counter
                Solution.Trie.counter += 1
        
        def getNode(self, curStr):
            cur = self
            for curChar in curStr:
                curId = ord(curChar) - ord('a')
                if not cur.child[curId]:
                    return None
                cur = cur.child[curId]
            return cur
        
        def getId(self, curStr):
            cur = self.getNode(curStr)
            return (cur.id if cur else -1)


    def minimumCost(self, source: str, target: str, original: List[str], changed: List[str], cost: List[int]) -> int:
        
        inf = 10 ** 15
        self.Trie.counter = 0
        root = self.Trie()

        for curStr in original:
            root.insert(curStr)
        for curStr in changed:
            root.insert(curStr)
        
        edgeMatrix = [[inf] * self.Trie.counter for _ in range(self.Trie.counter)]
        for i in range(self.Trie.counter):
            edgeMatrix[i][i] = 0
        for orgStr, tarStr, curCost in zip(original, changed, cost):
            u, v = root.getId(orgStr), root.getId(tarStr)
            edgeMatrix[u][v] = min(edgeMatrix[u][v], curCost)
        for k in range(self.Trie.counter):
            for i in range(self.Trie.counter):
                if edgeMatrix[i][k] == inf:
                    continue
                for j in range(self.Trie.counter):
                    if edgeMatrix[k][j] != inf:
                        edgeMatrix[i][j] = min(edgeMatrix[i][j], edgeMatrix[i][k] + edgeMatrix[k][j])

        memTable = [inf] * len(source)
        for i in range(len(source) - 1, -1, -1):
            if source[i] == target[i]:
                memTable[i] = min(memTable[i], 0 if i == len(source) - 1 else memTable[i + 1])
            curSrc, curTar = root, root
            for j in range(i, len(source)):
                srcChar, tarChar = source[j], target[j]
                srcId, tarId = ord(srcChar) - ord('a'), ord(tarChar) - ord('a')
                if not curSrc.child[srcId] or not curTar.child[tarId]:
                    break
                curSrc, curTar = curSrc.child[srcId], curTar.child[tarId]
                if curSrc.id != -1 and curTar.id != -1:
                    nextCost = 0 if j == len(source) - 1 else memTable[j + 1]
                    if edgeMatrix[curSrc.id][curTar.id] != inf and nextCost != inf:
                        memTable[i] = min(memTable[i], edgeMatrix[curSrc.id][curTar.id] + nextCost)
        
        return memTable[0] if memTable[0] != inf else -1

File: cpp/Minimum_Cost_to_Convert_String_II_2977.cpp
class Solution {
public:

    static constexpr long long inf = 1e15;

    struct Trie {

        static inline int counter = 0;

        Trie* child[26];
        int id;

        Trie() {
            for (int i = 0; i < 26; i++) child[i] = nullptr;
            id = -1;
        }

        void insert(string& curStr) {
            Trie* cur = this;
            for (auto& curChar: curStr) {
                int idx = curChar - 'a';
                if (!(cur->child)[idx]) {
                    (cur->child)[idx] = new Trie();
                }
                cur = (cur->child)[idx];
            }
            if (cur->id == -1) {
                cur->id = counter++;
            }
        }

        Trie* getNode(string& curStr) {
            Trie* cur = this;
            for (auto& curChar: curStr) {
                int idx = curChar - 'a';
                if (!(cur->child)[idx]) return nullptr;
                cur = (cur->child)[idx];
            }
            return cur;
        }

        int getId(string& curStr) {
            Trie* cur = getNode(curStr);
            return (!cur ? -1 : cur->id);
        }

        ~Trie() {
            for (int i = 0; i < 26; i++) {
                if (child[i]) {
                    delete child[i];
                    child[i] = nullptr;
                }
            }
        }

    };

    long long minimumCost(string source, string target, vector<string>& original, vector<string>& changed, vector<int>& cost) {

        Trie::counter = 0;
        Trie* root = new Trie();

        for (auto& strNode: original) {
            root->insert(strNode);
        }
        for (auto& strNode: changed) {
            root->insert(strNode);
        }

        vector <vector <long long>> edgeMatrix(Trie::counter, vector <long long> (Trie::counter, inf));
        for (int i = 0; i < Trie::counter; i++) {
            edgeMatrix[i][i] = 0;
        }
        for (int i = 0; i < original.size(); i++) {
            int u = root->getId(original[i]), v = root->getId(changed[i]);
            edgeMatrix[u][v] = min(edgeMatrix[u][v], static_cast<long long>(cost[i]));
        }
        for (int k = 0; k < Trie::counter; k++) {
            for (int i = 0; i < Trie::counter; i++) {
                if (edgeMatrix[i][k] == inf) continue;
                for (int j = 0; j < Trie::counter; j++) {
                    if (edgeMatrix[k][j] < inf) {
                        edgeMatrix[i][j] = min(edgeMatrix[i][j], edgeMatrix[i][k] + edgeMatrix[k][j]);
                    }
                }
            }
        }

        vector <long long> memTable(source.length(), inf);
        for (int i = source.length() - 1; i >= 0; i--) {
            long long ans = inf;
            if (source[i] == target[i]) {
                ans = min(ans, (i == source.length() - 1 ? 0LL : memTable[i + 1]));
            }
            Trie *curSrc = root, *curTar = root;
            for (int j = i; j < source.length(); j++) {
                int srcId = source[j] - 'a', tarId = target[j] - 'a';
                if (!(curSrc->child)[srcId] or !(curTar->child)[tarId]) break;
                curSrc = (curSrc->child)[srcId], curTar = (curTar->child)[tarId];
                if (curSrc->id != -1 and curTar->id != -1 and edgeMatrix[curSrc->id][curTar->id] != inf) {
                    long long next = (j == source.length() - 1 ? 0LL : memTable[j + 1]);
                    if (next != inf) {
                        ans = min(ans, edgeMatrix[curSrc->id][curTar->id] + next);
                    }
                }
            }
            memTable[i] = ans;
        }
        
        delete root;

        return (memTable[0] == inf ? -1LL : memTable[0]);

    }
};
